common/short-hex-str/src/lib.rs

```diff
@@ -81,22 +85,30 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
 /// Format a byte as hex. Returns a tuple containing the first character and then
 /// the second character as ASCII bytes.
 #[inline(always)]
-fn byte2hex(byte: u8) -> (u8, u8) {
-    let hi = HEX_CHARS_LOWER[((byte >> 4) & 0x0f) as usize];
+fn byte2hex(byte: u8) -> Result<(u8, u8), Error> {
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
```

Comment:
@terrytzh, for what it's worth, @bmwill pointed out that this overflow shouldn't ever be possible given the nature of `u8` and shift right. 

However, until we have some way of notifying the linter/arithmetic checker that we've manually reviewed this overflow and it's okay as-is (I'm not aware of a fine-grained way of notifying the arithmetic checker), let's leave it as is.

Reply:
Thanks @phlip9!

> surprisingly, the compiler is actually able to (almost) completely remove both the checked arithmetic and error paths (since they're impossible) and my local micro benches only have a ~12% regression here.

What regression are you referring to here? Compiler non-optimizations vs. optimizations, or something else?

Thanks for following up with the revert: https://github.com/diem/diem/pull/8509. Will trust you that these paths are safe 😄

Reply:
I have an old set of micro benchmarks sitting around for when I wrote short-hex-str. they were never published (they're a bit messy). I just added a new case with the error handling and checked arithmetic and compared the benches with optimizations etc. 😅